### ISSUE TITLE
JobAgent polling time delays

### ISSUE THREAD

**Issue by marianne013** [DIRACGrid/DIRAC#6116]
Background: We've been trying to work out why the JobAgent would take an increasing amount of time per cycle.
Looking at the pilot log of a completed job (htcondorce://ceprod02.grid.hep.ph.ic.ac.uk/378819.0 on the certification server) we see that in every cycle more of the errors below start showing up and each cycle is 4 min longer than the previous one, so even with 'only' 5 cycles in fill mode it takes a very long time to finish even if no more jobs are queued. 

```
2022-05-25 13:24:43 UTC WorkloadManagement/JobAgent DEBUG: Updating configuration internals
2022-05-25 13:24:43 UTC WorkloadManagement/JobAgent DEBUG: Fail to create Producer: Requested destination not found in the CS: Monitoring::Queues::dirac.monitoring
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent/DIRAC.Core.Tornado.Client.ClientSelector DEBUG: Trying to autodetect client for https://lbcertifdirac70.cern.ch:9135/Configuration/Server
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent/DIRAC.Core.Tornado.Client.ClientSelector VERBOSE: Using HTTPS for service https://lbcertifdirac70.cern.ch:9135/Configuration/Server
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent DEBUG: Already given a valid url https://lbcertifdirac70.cern.ch:9135/Configuration/Server
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent DEBUG: Already given a valid url https://lbcertifdirac70.cern.ch:9135/Configuration/Server
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent DEBUG: Updating configuration internals
2022-05-25 13:26:23 UTC WorkloadManagement/JobAgent DEBUG: Fail to create Producer: Requested destination not found in the CS: Monitoring::Queues::dirac.monitoring
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE: ----------------------------------------
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE: Agent module WorkloadManagement/JobAgent run summary
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Executed 10 times previously
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Cycle took 1200.36 seconds
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Average execution time: 558.88 seconds
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Polling time: 120 seconds
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Average execution/polling time: 465.73%
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE:  Cycle was successful
2022-05-25 13:27:52 UTC WorkloadManagement/JobAgent/WorkloadManagement/JobAgent NOTICE: ----------------------------------------

```

I assume this is unintended, but I have no idea how to fix it.
Related: #6106

**Comment by marianne013** [comment]
@fstagni  I think the monitoring error is a red herring. I just upgraded the pilot version on one of our preprod servers from v7.3.12 to v7.3.22 and the problem (increasing amount per cycle, up to 1200 s) has started appearing, but without the dirac.monitoring error. I don't think it was present in  v7.3.12, but I haven't run conclusive tests yet.

**Comment by fstagni** [comment]
The change in JobAgent that you mention was done in https://github.com/DIRACGrid/DIRAC/pull/5957 and IIRC connected to this change in Pilot: https://github.com/DIRACGrid/Pilot/pull/151/files where, in the case of PoolCE, the PollingTime is set to 20 seconds.

I think the issue is actually this:
> DIRAC submits slightly more pilots than jobs

Which is a strategy that works well for the case when all the following are met:
- the pilots are "disposable" (which is less the case for VMs than for "standard" pilots)
- the pilots do not try and match more than a job at a single run (basically when FillingMode is off, or when there's one cycle of the JobAgent)
- there are always jobs to match

A quick-and-not-so-good-fix is to set the PollingTime of the JobAgent to 20s, and this could be done now, and would not hurt. 
But most probably the strategy of how many pilots/VMs are submitted should be adjusted -- but this is a second not-so-easy-fix, worth a little discussion.

**Comment by marianne013** [comment]
@fstagni To get back to this, we (GridPP) are still trying to understand what was the reasoning behind having a pilot literally sit around for hours. I don't think there's a reason not to wait let's say 20 min, but as you say pilots are disposable and there will always be more pilots. If nothing has happened in 20 min, there are probably no jobs for that VO around.
The over submitting is definitely helpful, as especially in a multi-VO environment DIRAC cannot reliably predict which site will actually run the work loads faster and that has saved us more than once.  However sites get really really grouchy when you send them empty pilots and then don't get rid of them fast enough.

**Comment by fstagni** [comment]
> If nothing has happened in 20 min, there are probably no jobs for that VO around.

Even 10 or even 5 minutes. But I would still do this:

> A quick-and-not-so-good-fix is to set the PollingTime of the JobAgent to 20s, and this could be done now, and would not hurt.

**Comment by marianne013** [comment]
You still haven't told me what the reasoning was :-D

**Comment by fstagni** [comment]
The original change (buggy) was done in https://github.com/DIRACGrid/DIRAC/pull/5001, then corrected in https://github.com/DIRACGrid/DIRAC/pull/5957. TBH, atm I can't remember the original use case.

**Comment by marianne013** [comment]
How about "Maybe we made a mistake when implementing this feature, so we do one retry and then give up" instead ?

**Comment by fstagni** [comment]
Not the most complex fix, but #6469 + https://github.com/DIRACGrid/Pilot/pull/164 are the simplest approximations.
